File: code/backend/src/modules/clientes/clienteService.ts
import { ClienteRepository } from "./clienteRepository.js";

interface CriarClienteInput {
  nome: string;
  telefone?: string;
  userId: number;
}

interface AtualizarClienteInput {
  nome?: string;
  telefone?: string;
}

export class ClienteService {
  private clienteRepository: ClienteRepository;

  constructor() {
    this.clienteRepository = new ClienteRepository();
  }

  async criar(data: CriarClienteInput) {
    return this.clienteRepository.criar(data);
  }

  async listarTodos() {
    return this.clienteRepository.listarTodos();
  }

  async buscarPorId(id: number) {
    const cliente = await this.clienteRepository.buscarPorId(id);

    if (!cliente) {
      throw new Error("Cliente nao encontrado.");
    }

    return cliente;
  }

  async atualizar(id: number, data: AtualizarClienteInput) {
    await this.buscarPorId(id);
    return this.clienteRepository.atualizar(id, data);
  }

  async deletar(id: number) {
    await this.buscarPorId(id);
    return this.clienteRepository.deletar(id);
  }
}

File: code/mist_mobile/lib/services/cliente_service.dart
import '/models/cliente.dart';
import '/services/api_service.dart';

class ClienteService {
  static Future<void> criar({
    required String nome,
    required String email,
    required String telefone,
  }) async {
    final body = {
      'nome': nome,
      'email': email,
      'telefone': telefone,
    };

    await ApiService.post('/clientes', body);
  }

  static Future<List<Cliente>> listarTodos() async {
    final response = await ApiService.get('/clientes');
    
    if (response is List) {
      return response.map((e) => Cliente.fromJson(e)).toList();
    }
    
    return [];
  }

  static Future<Cliente> buscarPorId(int id) async {
    final response = await ApiService.get('/clientes/$id');
    return Cliente.fromJson(response);
  }

  static Future<void> atualizar({
    required int id,
    String? nome,
    String? email,
    String? telefone,
  }) async {
    final body = <String, dynamic>{
      if (nome != null) 'nome': nome,
      if (email != null) 'email': email,
      if (telefone != null) 'telefone': telefone,
    };

    await ApiService.put('/clientes/$id', body);
  }

  static Future<void> deletar(int id) async {
    await ApiService.delete('/clientes/$id');
  }
}
